### ISSUE TITLE
Add Microsoft BitNet b1.58 support (I2_S ternary, BitNetTransformerModel, config-driven activation)

### ISSUE THREAD

**Issue by jamesburton** [jamesburton/dotLLM#77]
## Summary

Add support for **Microsoft BitNet b1.58** — natively-trained ternary-weight ({-1, 0, +1}, ≈1.58 bits) LLMs — to dotLLM's CPU backend, matching Microsoft's official deployment by loading the **I2_S** GGUF quantization used in `microsoft/bitnet-b1.58-2B-4T-gguf`.

Two design directives for this issue:
1. Implement a dedicated **`BitNetTransformerModel`** (selected by the architecture factory for `Architecture.BitNet`).
2. Make the **FFN activation function config-driven** (it is currently hardcoded to SwiGLU/SiLU in the forward pass).

## Target model

| | |
|---|---|
| Primary | [`microsoft/bitnet-b1.58-2B-4T-gguf`](https://huggingface.co/microsoft/bitnet-b1.58-2B-4T-gguf) — `ggml-model-i2_s.gguf` (~1.19 GB) |
| Reference (math) | [`microsoft/bitnet-b1.58-2B-4T`](https://huggingface.co/microsoft/bitnet-b1.58-2B-4T) (transformers `BitNetForCausalLM`, `model_type: bitnet`) |
| License | MIT |
| Footprint | ~1.2 GB → runs comfortably on T1 (RTX 3060 12 GB) and CPU-only |

### Architecture (from `config.json`)

| Field | Value |
|---|---|
| hidden_size | 2560 |
| intermediate_size | 6912 |
| num_hidden_layers | 30 |
| num_attention_heads / kv | 20 / 5 (GQA, head_dim 128) |
| vocab_size | 128256 (LLaMA-3 tokenizer) |
| max_position_embeddings | 4096 |
| hidden_act | **`relu2`** (squared ReLU) |
| rms_norm_eps | 1e-5 |
| rope_theta | 500000.0 |
| tie_word_embeddings | true |
| bias | none |

## What's different from a standard Llama block

Confirmed against the reference modeling code (`transformers/models/bitnet/modeling_bitnet.py`):

1. **Two extra sub-norms per layer** (this is the "Sub-LN" of the BitNet paper):
   - `attn_sub_norm`: `RMSNorm(hidden)` applied to the **attention output before `o_proj`**.
   - `ffn_sub_norm`: `RMSNorm(intermediate)` applied to the **gated FFN intermediate before `down_proj`**.
   ```python
   # BitNetMLP.forward
   down_proj(self.ffn_sub_norm( act_fn(gate_proj(x)) * up_proj(x) ))
   # BitNetAttention: o_proj(attn_sub_norm(attn_output))
   ```
2. **Activation is squared ReLU (ReLU²)**, not SiLU — applied to the gate branch of a gated FFN: `relu(gate)² * up`.
3. **Ternary weights via I2_S** on all attention + FFN projections (BitLinear).
4. **RoPE** (theta 500000) — *not* ALiBi. Reuses the existing RoPE path; no new position encoding needed.
5. No biases anywhere; tied embeddings (both already handled by the loader).

### I2_S quantization format (bitnet.cpp)

- Block `QK_I2_S = 128` (x86) / `64` (ARM); **4 ternary codes packed per byte (2 bits each)**.
- Code → value: `0 → -1`, `1 → 0`, `2 → +1` (i.e. `value = code - 1`).
- Byte packing within a 128-element block: 32 bytes; byte `group_pos` (0..31) holds positions `{j, j+32, j+64, j+96}` at bit offsets `6,4,2,0` respectively.
- **Per-tensor float scale** stored as a single `float` immediately after the packed data (offset `n/4` bytes).
- **Action item:** confirm the exact `ggml_type` id and the GGUF tensor names (`blk.N.attn_sub_norm.weight`, `blk.N.ffn_sub_norm.weight`, etc.) by dumping the actual gguf metadata as the first implementation step — the file is the source of truth (bitnet.cpp's I2_S is not a mainline ggml type).

## Proposed implementation

The codebase is already strongly parameterized and uses an **optional-weight-via-null-check** idiom (QK-norms, biases). The two sub-norms slot into that exact pattern; the activation becomes a config-driven dispatch. `BitNetTransformerModel` is a thin subclass wiring BitNet specifics; the shared forward pass gains the sub-norm + activation hooks gated on config/weight presence (keeps the "single parameterized TransformerBlock" philosophy).

**Core / config**
- `DotLLM.Core/Configuration/QuantizationType.cs` — add `I2_S` (value confirmed from gguf).
- `DotLLM.Core/Configuration/QuantizationTypeExtensions.cs` — byte-size: `n/4 + sizeof(float)` per tensor (per-tensor scale).
- `DotLLM.Core/Configuration/Architecture.cs` — add `BitNet`.
- `DotLLM.Core/Configuration/ActivationFunction.cs` — add `ReLU2`.
- `DotLLM.Core/Models/ModelConfig.cs` — already carries `ActivationFunction`; ensure it is populated and consumed.

**Loader / config extraction**
- `GgufModelConfigExtractor.cs` — map arch `"bitnet"` → `Architecture.BitNet`; set `ActivationFunction.ReLU2` for BitNet; RoPE already extracted from `bitnet.rope.freq_base`.
- `TransformerWeights.cs` — load optional `attn_sub_norm` / `ffn_sub_norm` via the existing `LoadOptionalNorm` helper (null for non-BitNet); add the two fields to `TransformerLayerWeights`.

**CPU kernels**
- `DotLLM.Cpu/Kernels/Dequantize.cs` — add `DequantizeI2_S` (scalar reference first, then AVX2).
- `DotLLM.Cpu/Kernels/MatMul.cs` — ternary BitLinear GEMV/GEMM. **Phase 1: route I2_S through the existing `GemvDequantFallback`/`GemmDequantFallback`** (already wired via the `else` branch) for a correctness-first path; **Phase 2:** dedicated ternary kernel (int8 activations × ternary, per-tensor rescale) + R4 repacking support.
- New `DotLLM.Cpu/Kernels/ReLu2.cs` (or extend `FusedOps`) — `relu(gate)² * up` fused gated activation.

**Forward pass** (`TransformerModel.cs`)
- After attention, before O-projection quantize: if `AttnSubNormWeight != null`, apply RMSNorm to `attnOut`.
- After the gated activation, before Down-projection quantize: if `FfnSubNormWeight != null`, apply RMSNorm over `intermediateSize`.
- Replace the hardcoded `FusedOps.SwiGLU` call with a dispatch on `config.ActivationFunction` (SiLU/SwiGLU today, ReLU² for BitNet).

**Architecture wiring**
- `TransformerArchitecture.cs` — keep as-is for Llama family.
- New `BitNetTransformerModel` + register `Architecture.BitNet` in the factory dispatch (`ModelLoader`).

## Future acceleration & testing (out of scope for first PR, noted for roadmap)

- **CUDA GEMV**: dotLLM's CUDA backend has GEMV kernels available; a W2A8 (2-bit weight × int8 activation) ternary GEMV is a natural follow-up for GPU acceleration. bitnet.cpp shipped official GPU kernels (May 2025) as a reference.
- **AMD / Strix Halo (T2)**: ROCm / Vulkan path on the Strix Halo workstation (128 GB unified) should show strong performance on the memory-bandwidth-bound ternary kernels — good target for larger BitNet variants and long context.
- **Intel iGPU**: framework can assist with iGPU development/testing at a later stage on Intel Core Ultra 7.

## Acceptance criteria

- [ ] `microsoft/bitnet-b1.58-2B-4T-gguf` (`ggml-model-i2_s.gguf`) loads without error; config extracted correctly (arch BitNet, relu2, RoPE theta 500000).
- [ ] I2_S dequant matches a reference vector (unit test, scalar + SIMD agree).
- [ ] ReLU² activation matches reference (unit test).
- [ ] Sub-norm placement verified against `modeling_bitnet.py` math.
- [ ] End-to-end generation produces coherent text (integration test / manual sanity check).
- [ ] Zero allocations on the inference hot path; no regression for existing architectures.

## TDD / validation plan

Per CLAUDE.md ("verify numerical accuracy against scalar reference first, then optimize"):
1. I2_S unpack/dequant unit tests (known byte patterns → expected ternary × scale).
2. ReLU² unit test vs scalar reference.
3. Correctness-first forward via dequant fallback; compare a few logits/next-token against the reference where feasible.
4. Only then add the optimized ternary kernel + benchmark vs the fallback.

## References

- Models: [b1.58-2B-4T](https://huggingface.co/microsoft/bitnet-b1.58-2B-4T) · [-gguf](https://huggingface.co/microsoft/bitnet-b1.58-2B-4T-gguf) · [-bf16](https://huggingface.co/microsoft/bitnet-b1.58-2B-4T-bf16)
- Papers: arXiv **2504.12285** (b1.58 2B4T Technical Report) · **2402.17764** (Era of 1-bit LLMs) · **2310.11453** (original BitNet) · **2502.11880** (bitnet.cpp)
- Code: [microsoft/BitNet](https://github.com/microsoft/BitNet) · [llama.cpp ternary PR #8151](https://github.com/ggml-org/llama.cpp/pull/8151)
- Docs to read: `docs/QUANTIZATION.md`, `docs/MODEL_CONFIG.md`, `docs/GGUF_FORMAT.md`

**Comment by jamesburton** [comment]
Hardware caveat correcting the CPU-optimization notes above: the CPU dev workstation is a dual Xeon X5670 (Westmere, 2010) — **SSE4.2 only, no AVX/AVX2/FMA** (`Avx2.IsSupported == False` at runtime). Two consequences:

- Our W2A8/int8 CPU kernels gate on `Avx2.IsSupported`, so on this box they never executed — they fell back to the SSE `TensorPrimitives.Dot` float path. The earlier "W2A8 micro-ops do not beat the parallel-float kernel" result was the **SSE float path measured twice**; the int8/maddubs/VNNI path is **untested on this hardware** and is expected to help on an AVX2/VNNI CPU. The shipped CPU kernel (parallel float) is the right choice *for Westmere*.
- The ~1.87 tok/s decode / ~8.67 prefill numbers are a 15-year-old Westmere ceiling, not a modern-CPU figure. **bitnet.cpp cannot run here** (its x86 I2_S/TL2 kernels are AVX2-only).

→ The CPU comparison vs bitnet.cpp and validation of our AVX2/W2A8 path are deferred to AVX2 hardware (Intel Core Ultra 7 Meteor Lake / Strix Halo Zen5). On this box, the GPU (RTX 3060) is the productive path.

**Comment by jamesburton** [comment]
BitNet b1.58 support is complete and merged to `dev`.

CPU + CUDA BitNet (I2_S ternary), config-driven ReLU² activation, Sub-LN, and BitNet-LoRA (M1 CPU / M2 GPU) all landed on `dev` via **PR #81** and follow-ups **#84** (hybrid LoRA) / **#85** (hybrid+BitNet I2_S fix) / **#90** (dtype-aware GPU adapter staging). Verified end-to-end: coherent generation, CPU↔GPU LoRA parity (cosine 0.999995), and a real LoRA trained on BitNet itself and served on the I2_S GGUF.

Offered upstream as **kkokosa/dotLLM#334** (issue-first; PR on request). PR #78 (into the fork's `main`) closed — that route isn't wanted.

Closing as done.
